src/backend/lib/upgrade-manager.js

```diff
@@ -1,192 +1,275 @@
 
-  stopServices() {
-    log("got request to stop services");
-    this.server.drain();
-    this.downloader.stop();
+      /** @type {TransferProgress} */
+      const download = {
+        id: candidateHash,
+        sofar: 0,
+        total: upgradeCandidate.size,
+      };
 
-    if (this.upgradeSearchTimeout) {
-      clearTimeout(this.upgradeSearchTimeout);
+      const rs = this._discovery.createReadStream(candidateHash);
+      const progress = progressStream(
+        { length: upgradeCandidate.size },
+        progress => {
+          // Warning: we mutate the download object to avoid recreating an
+          // object on every progress event, and also re-setting this.#downloads
+          download.sofar = progress.transferred;
+          this.throttledEmitState();
+        }
+      );
+      const ws = this._storage.createWriteStream({ hash: candidateHash });
+
+      stream.pipeline(rs, progress, ws, e => {
+        this.#downloads.delete(candidateHash);
+      });
+
+      this.#downloads.set(candidateHash, {
+        stream: rs,
+        progress: download,
+        installerInfo: upgradeCandidate,
+      });
+    } catch (e) {
+      log("Uncaught error processing available downloads", e);
+      this.setState({ value: "error", error: e });
     }
   }
 
+  /**
+   * Get current state of UpgradeManager
+   *
+   * @returns {UpgradeState}
+   */
   getState() {
-    log("got request for state");
-    this.emitState();
+    return {
+      ...this.#state,
+      // Don't like this too much, but for now is better than duplicating
+      // downloads on state
+      downloads: Array.from(this.#downloads.values()).map(
+        download => download.progress
+      ),
+    };
   }
 
-  emitState() {
-    const result = validate(this.state);
-    if (result) {
-      const msg = result.join(",");
-      this.emit("error", { message: msg });
-    } else {
-      this.sanitizeState();
-      this.emit("state", this.state);
-    }
+  /**
+   * Set state of UpgradeManager, **merges with current state**. If you pass
+   * state with value: "error" you _must_ also pass an error property
+   *
+   * @param {Partial<Exclude<UpgradeState, { value: "error", error: Error }>> | { value: "error", error: Error }} statePartial
+   */
+  setState(statePartial) {
+    this.#state = { ...this.#state, ...statePartial };
+    this.throttledEmitState();
   }
 
-  setApkInfo(apkPath, version, cb) {
-    log("setApkInfo", apkPath, version);
-    this.storage.setApkInfo(apkPath, version, cb);
+  /**
+   * Start the downloader services.
+   *
+   * 1. Starts the server on an available port
+   * 2. Starts the download discovery service
+   * 3. Starts announcing on port
+   *
+   * @returns {Promise<void>} Resolves when started
+   */
+  async _start() {
+    log("got request to start services");
+    const port = await getPort();
+    await Promise.all([this._discovery.start(port), this._server.start(port)]);
+    log("Started services");
```

Comment:
Why does `_stop` need to call `setState` but `_start` doesn't?

Reply:
Good catch! That is a bug, it should not be there in `_stop`, not sure how it sneaked in!